aws/resource_aws_wafv2_web_acl.go

```diff
@@ -0,0 +1,843 @@
+	}
+
+	arn := aws.StringValue(resp.WebACL.ARN)
+	tags, err := keyvaluetags.Wafv2ListTags(conn, arn)
+	if err != nil {
+		return fmt.Errorf("Error listing tags for WAFv2 WebACL (%s): %s", arn, err)
+	}
+
+	if err := d.Set("tags", tags.IgnoreAws().IgnoreConfig(ignoreTagsConfig).Map()); err != nil {
+		return fmt.Errorf("Error setting tags: %s", err)
+	}
+
+	return nil
+}
+
+func resourceAwsWafv2WebACLUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).wafv2conn
+
+	log.Printf("[INFO] Updating WAFv2 WebACL %s", d.Id())
+
+	u := &wafv2.UpdateWebACLInput{
+		Id:               aws.String(d.Id()),
+		Name:             aws.String(d.Get("name").(string)),
+		Scope:            aws.String(d.Get("scope").(string)),
+		LockToken:        aws.String(d.Get("lock_token").(string)),
+		DefaultAction:    expandWafv2DefaultAction(d.Get("default_action").([]interface{})),
+		Rules:            expandWafv2Rules(d.Get("rule").(*schema.Set).List()),
+		VisibilityConfig: expandWafv2VisibilityConfig(d.Get("visibility_config").([]interface{})),
+	}
+
+	if v, ok := d.GetOk("description"); ok && len(v.(string)) > 0 {
+		u.Description = aws.String(v.(string))
+	}
+
+	err := resource.Retry(5*time.Minute, func() *resource.RetryError {
+		_, err := conn.UpdateWebACL(u)
+		if err != nil {
+			if isAWSErr(err, wafv2.ErrCodeWAFUnavailableEntityException, "") {
+				return resource.RetryableError(err)
+			}
+			return resource.NonRetryableError(err)
+		}
+		return nil
+	})
+
+	if isResourceTimeoutError(err) {
+		_, err = conn.UpdateWebACL(u)
+	}
+
+	if err != nil {
+		return fmt.Errorf("Error updating WAFv2 WebACL: %s", err)
+	}
+
+	if d.HasChange("tags") {
+		o, n := d.GetChange("tags")
+		if err := keyvaluetags.Wafv2UpdateTags(conn, d.Get("arn").(string), o, n); err != nil {
+			return fmt.Errorf("error updating tags: %s", err)
+		}
+	}
+
+	return resourceAwsWafv2WebACLRead(d, meta)
+}
+
+func resourceAwsWafv2WebACLDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).wafv2conn
+
+	log.Printf("[INFO] Deleting WAFv2 WebACL %s", d.Id())
+
+	r := &wafv2.DeleteWebACLInput{
+		Id:        aws.String(d.Id()),
+		Name:      aws.String(d.Get("name").(string)),
+		Scope:     aws.String(d.Get("scope").(string)),
+		LockToken: aws.String(d.Get("lock_token").(string)),
+	}
+
+	err := resource.Retry(5*time.Minute, func() *resource.RetryError {
+		_, err := conn.DeleteWebACL(r)
+		if err != nil {
+			if isAWSErr(err, wafv2.ErrCodeWAFAssociatedItemException, "") {
+				return resource.RetryableError(err)
+			}
+			if isAWSErr(err, wafv2.ErrCodeWAFUnavailableEntityException, "") {
+				return resource.RetryableError(err)
+			}
+			return resource.NonRetryableError(err)
+		}
+		return nil
+	})
+
+	if isResourceTimeoutError(err) {
+		_, err = conn.DeleteWebACL(r)
+	}
+
+	if err != nil {
+		return fmt.Errorf("Error deleting WAFv2 WebACL: %s", err)
+	}
+
+	return nil
+}
+
+func wafv2WebACLRootStatementSchema(level int) *schema.Schema {
```

Comment:
@pvanbuijtene latest discussion in the RuleGroup resource issue #1175 revealed the `statement` field in the RuleGroup resource doesn't include `managed_rules` as included in this schema definition; this shouldn't block the release of this PR but just noting here that this function may later need to be renamed to `wafv2RootStatementSchema` and replace the current function in the wafv2_helper file as the `statement` field in both resources look identical per AWS docs

Reply:
@anGie44 AFAIK managed rules can only be added on a Web ACL, the docs are not 100% correct 😬 ... this is also why I had to create a different root statement function.

Reply:
Sadly I encountered more undocumented limitations like this while working on these wafv2 resources

Reply:
ah ok I had a feeling there was some backgrounf behind the design decision (and looking with AWS Console I do only see the managed rule group option within the Web ACL UI)